ocaml/lambda/value_rec_compiler.ml

```diff
@@ -0,0 +1,822 @@
+}
+
+let dynamic_size () =
+  Misc.fatal_error "letrec: No size found for Static binding"
+
+(* [join_sizes] is used to compute the size of an expression with multiple
+   branches. Such expressions are normally classified as [Dynamic] by
+   [Value_rec_check], so the default behaviour is a compile-time failure.
+   However, for partial pattern-matching (typically in let bindings)
+   the compiler will later add a branch for the failing cases, and this
+   is handled here with the [Unreachable] case.
+   Note that the current compilation scheme would work if we allowed the
+   [Constant] and [Block] cases to be joined, but [Function] needs to be
+   a single function. *)
+let join_sizes size1 size2 =
+  match size1, size2 with
+  | Unreachable, size | size, Unreachable -> size
+  | _, _ -> dynamic_size ()
+
+let compute_static_size lam =
+  let rec compute_expression_size env lam =
+    match lam with
+    | Lvar v ->
+      begin match Ident.Map.find_opt v env with
+      | None ->
+        dynamic_size ()
+      | Some binding_size ->
+        Lazy_backtrack.force
+          (fun { lambda; env } -> compute_expression_size env lambda)
+          binding_size
+      end
+    | Lmutvar _ -> dynamic_size ()
+    | Lconst _ -> Constant
+    | Lapply _ -> dynamic_size ()
+    | Lfunction _ -> Function
+    | Llet (_, _, id, def, body) ->
+      let env =
+        Ident.Map.add id (Lazy_backtrack.create { lambda = def; env }) env
+      in
+      compute_expression_size env body
+    | Lmutlet(_, _, _, body) ->
+      compute_expression_size env body
+    | Lletrec (bindings, body) ->
+      let env =
+        List.fold_left (fun env_acc { id; def = _ } ->
+            Ident.Map.add id (Lazy_backtrack.create_forced Function) env_acc)
+          env bindings
+      in
+      compute_expression_size env body
+    | Lprim (p, args, _) ->
+      size_of_primitive env p args
+    | Lswitch (_, sw, _, _) ->
+      let fail_case =
+        match sw.sw_failaction with
+        | None -> []
+        | Some fail -> [0 (* ignored *), fail]
+      in
+      compute_and_join_sizes_switch env [sw.sw_consts; sw.sw_blocks; fail_case]
+    | Lstringswitch (_, cases, fail, _, _) ->
+      let fail_case =
+        match fail with
+        | None -> []
+        | Some fail -> ["" (* ignored *), fail]
+      in
+      compute_and_join_sizes_switch env [cases; fail_case]
+    | Lstaticraise _ -> Unreachable
+    | Lstaticcatch (body, _, handler, _)
+    | Ltrywith (body, _, handler, _) ->
+      compute_and_join_sizes env [body; handler]
+    | Lifthenelse (_cond, ifso, ifnot, _) ->
+      compute_and_join_sizes env [ifso; ifnot]
+    | Lsequence (_, e) ->
+      compute_expression_size env e
+    | Lwhile _
+    | Lfor _
+    | Lassign _ -> Constant
+    | Lsend _ -> dynamic_size ()
+    | Levent (e, _) ->
+      compute_expression_size env e
+    | Lifused _ -> Constant
+    | Lregion (e, _) ->
+      compute_expression_size env e
+    | Lexclave e ->
```

Comment:
I'm not familiar enough with the `Lexclave` constructor to know which contexts it is allowed in. If you're confident that such an expression will never appear anywhere in a recursive definition, I can put `dynamic_size` instead (which raises `Misc.Fatal_error` instead of `Assertion_failure`).

Reply:
"never appear anywhere": It could be in a function, like `let rec f x = exclave_ ...`. But there can't be a `let rec x = exclave_ ...`.